Fix for 'rsync_cleanup_after_native_cp_al() only works on directories' fail when sync_first on and cmd_cp not set (#133).

sgpinkus opened the pull request:
Refs #133. Solves same issue as #177. IMO better than #177 since it's clear why ending slashes are needed and they are only added exactly where they are needed. Also adds includes a test case.

DrHyde commented on t/native_cp_al/conf/native_cp_al.conf.in:
```diff
@@ -0,0 +1,7 @@
+sync_first	1
+link_dest	0
+interval		hourly	6
+backup			@TEMP@	native_cp_al/
```
We're testing that the / get automatically applied aren't we? So it should be missing in the config here

sgpinkus replied:
That doesn't actually effect the failure mode. No harm in removing it. I will do that.

DrHyde commented on t/native_cp_al/conf/native_cp_al_link_dest.conf.in:
```diff
@@ -0,0 +1,7 @@
+sync_first	1
+link_dest	1
+interval		hourly	6
+backup			@TEMP@	native_cp_al/
```
And this / should be missing too I think

sgpinkus replied:
Just going to rm that test completely. See other comment about it.

DrHyde commented on t/native_cp_al/native_cp_al.t.in:
```diff
@@ -0,0 +1,39 @@
+	" sync failed");
+ok(-d "@SNAP@/.sync/native_cp_al",
+	" sync failed");
+ok(0 == rsnapshot("-c @TEST@/native_cp_al/conf/native_cp_al.conf hourly 2>&1"),
```
This is the only test that fails without your patch, is that what you expected?

DrHyde commented on t/native_cp_al/native_cp_al.t.in:
```diff
@@ -0,0 +1,39 @@
+	" sync failed");
+ok(-d "@SNAP@/.sync/native_cp_al",
+	" sync failed");
+ok(0 == rsnapshot("-c @TEST@/native_cp_al/conf/native_cp_al_link_dest.conf hourly 2>&1"),
```
I would expect this to fail as well without your patch

sgpinkus replied:
Nope. If `link_dest` is set the  `cp_al()` sub is not used in backup process and `cp_al()` is the only place that calls `rsync_cleanup_after_native_cp_al()` sub and only when `cp_al` is not set. On reflection I will just remove this test. It just causes confusion. Not sure why I added it in the first place.

DrHyde commented on t/native_cp_al/native_cp_al.t.in:
```diff
@@ -0,0 +1,39 @@
+ok(0 == rsnapshot("-c @TEST@/native_cp_al/conf/native_cp_al.conf hourly 2>&1"),
+	" hourly backup after sync failed");
+ok(-d "@SNAP@/hourly.0/native_cp_al",
+	" hourly backup failed");
```
I'm confused!

sgpinkus replied:
Haha :/ It's been a while since I've written a Perl test case. I think I got my conventions mixed up. The text message should describe what happens in case of success right? OK. I review some of the other tests and they ain't bothering putting any description in there so I'll do the same.

DrHyde replied:
Yes, the text normally describes what *should* have happened.

DrHyde commented on t/native_cp_al/native_cp_al.t.in:
```diff
@@ -0,0 +1,39 @@
+	" sync failed");
+ok(0 == rsnapshot("-c @TEST@/native_cp_al/conf/native_cp_al_link_dest.conf hourly 2>&1"),
+	" hourly backup after sync failed");
+ok(-d "@SNAP@/hourly.0/native_cp_al",
```
And this should say "succeeded" too?

DrHyde commented on t/native_cp_al/native_cp_al.t.in:
```diff
@@ -0,0 +1,39 @@
+	" snapshot root does not exist before testing starts");
+
+ok(0 == rsnapshot("-c @TEST@/native_cp_al/conf/native_cp_al.conf sync 2>&1"),
+	" sync failed");
```
An exit code of 0 means "success", note failure!

sgpinkus replied:
:+1:

DrHyde commented on t/native_cp_al/native_cp_al.t.in:
```diff
@@ -0,0 +1,39 @@
+ok(0 == rsnapshot("-c @TEST@/native_cp_al/conf/native_cp_al.conf sync 2>&1"),
+	" sync failed");
+ok(-d "@SNAP@/.sync/native_cp_al",
+	" sync failed");
```
Succeeded, I think?

DrHyde commented on t/native_cp_al/native_cp_al.t.in:
```diff
@@ -0,0 +1,39 @@
+ok(! remove_snapshot_root(),
+	" snapshot root does not exist before testing starts");
+
+ok(0 == rsnapshot("-c @TEST@/native_cp_al/conf/native_cp_al_link_dest.conf sync 2>&1"),,
```
Again, 0 == success exit code

DrHyde commented:
The code change looks correct, I'm just really confused by the tests!

sgpinkus commented:
> The code change looks correct, I'm just really confused by the tests!

np I can see that. I fixed all that, squashed commit and pushed back to this PR.  Hopefully the test makes more sense now,

DrHyde commented:
Thanks, merged
